File: src/filters.py
"""Пост-фильтры, которых нет (или не хватает) в самом API hh.ru:
- min_employer_rating — рейтинг компании (есть только в shards-ответе)
- exclude_keywords — стоп-слова в названии
- keyword в названии/описании — дублируем локально для надёжности
"""


def apply_filters(vacancies, search):
    min_rating = float(search.get("min_employer_rating") or 0.0)
    excludes = [w.lower() for w in (search.get("exclude_keywords") or [])]
    sf = (search.get("search_field") or "everywhere").strip()
    text = (search.get("text") or "").strip().lower()
    words = [w for w in text.split() if len(w) > 2] if text else []

    out = []
    for v in vacancies:
        # 1. Рейтинг работодателя
        if min_rating and (v.get("rating") or 0.0) < min_rating:
            # вакансии без рейтинга (0.0) при включённом пороге отсекаем
            continue
        name_l = (v.get("name") or "").lower()
        # 2. Стоп-слова
        if excludes and any(w in name_l for w in excludes):
            continue
        # 3. Локальная проверка ключевого слова (если искали строго в названии)
        if sf == "name" and words:
            if not all(w in name_l for w in words):
                continue
        out.append(v)
    return out


def salary_text(v):
    f, t, c = v.get("salary_from"), v.get("salary_to"), v.get("currency") or ""
    cur = {"RUR": "₽", "RUB": "₽", "KZT": "₸", "USD": "$", "EUR": "€"}.get(c, c)
    if f and t:
        return f"{f:,}–{t:,} {cur}".replace(",", " ")
    if f:
        return f"от {f:,} {cur}".replace(",", " ")
    if t:
        return f"до {t:,} {cur}".replace(",", " ")
    return "зарплата не указана"


SCHEDULE_RU = {
    "remote": "Удалённо",
    "fullDay": "Полный день",
    "shift": "Сменный",
    "flexible": "Гибкий",
    "flyInFlyOut": "Вахта",
}

File: netlify/lib/filters.mjs
// Пост-фильтры + форматирование. Порт src/filters.py.

export function applyFilters(vacancies, s) {
  const minRating = parseFloat(s.min_employer_rating || 0);
  const excludes = (s.exclude_keywords || []).map((w) => w.toLowerCase());
  const sf = (s.search_field || "everywhere").trim();
  const words = (s.text || "").trim().toLowerCase().split(/\s+/).filter((w) => w.length > 2);
  return vacancies.filter((v) => {
    if (minRating && (v.rating || 0) < minRating) return false;
    const name = (v.name || "").toLowerCase();
    if (excludes.length && excludes.some((w) => name.includes(w))) return false;
    if (sf === "name" && words.length && !words.every((w) => name.includes(w))) return false;
    return true;
  });
}

const CUR = { RUR: "₽", RUB: "₽", KZT: "₸", USD: "$", EUR: "€" };

export function salaryText(v) {
  const cur = CUR[v.currency] || v.currency || "";
  const fmt = (n) => n.toLocaleString("ru-RU");
  if (v.salary_from && v.salary_to) return `${fmt(v.salary_from)}–${fmt(v.salary_to)} ${cur}`;
  if (v.salary_from) return `от ${fmt(v.salary_from)} ${cur}`;
  if (v.salary_to) return `до ${fmt(v.salary_to)} ${cur}`;
  return "зарплата не указана";
}

export const SCHEDULE_RU = {
  remote: "Удалённо", fullDay: "Полный день", shift: "Сменный",
  flexible: "Гибкий", flyInFlyOut: "Вахта",
};

const esc = (s) => String(s ?? "").replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");

export function formatVacancy(v, searchNames = "") {
  const sched = SCHEDULE_RU[v.schedule] || v.schedule || "";
  const rating = v.rating
    ? `⭐ ${Number(v.rating).toFixed(1)} (${v.reviews_count || 0})` : "без рейтинга";
  const lines = [
    `💼 <b>${esc(v.name)}</b>`,
    `🏢 ${esc(v.employer)} · ${rating}`,
    `📍 ${esc(v.area)}${sched ? ` · ${sched}` : ""}`,
    `💰 ${salaryText(v)}`,
  ];
  if (searchNames) lines.push(`🔎 <i>${esc(searchNames)}</i>`);
  if (v.snippet) lines.push(`\n${esc(v.snippet.slice(0, 300))}`);
  lines.push(`🔗 ${v.url}`);
  return lines.join("\n");
}
